File: javascript/intro/rpg.py


    
    @value.setter
    def value(self, value_tuple): 
        gold, silver, copper = value_tuple
        self.__gold = gold
        self.__silver = silver
        self.__copper = copper

    def add(self, gold, silver, copper):
        self.__gold += gold
        self.__silver += silver
        self.__copper += copper

    def __add__(self,other): 
        return Currency(
            gold=self.__gold + other.__gold,
            silver=self.__silver + other.__silver,
            copper=self.__copper + other.__copper
        )

    def __iadd__(self, other): # __iadd__ means += 
        if not isinstance(other, Currency): # if not a instance of Currency
            raise TypeError('Can only add Currency to Currency') # stops other from being anything
        
        return Currency(
            gold=self.__gold + other.__gold,
            silver=self.__silver + other.__silver,
            copper=self.__copper + other.__copper
        )

    def __str__(self):
        return f'{self.__gold}G {self.__silver}S {self.__copper}C'


class Mage(Character):
    
    def __init__(self, name, race, health, attack,mana): 
        super().__init__(name, race, health, attack)
        self.mana = mana

    def battle(self, other):
        print(f'{self._name} casts a wicked spell {other._name}')

    def portal(self, destination):
        print(f'{self._name} opens a portal to {destination}')

class Burglar(Character):
    def battle(self, other):
        print(f'{self._name} sneaks in a stealth attack {other._name}')
class Warlock(Mage):
    pass

class Chest:
    def __init__(self, items, gold, silver, copper):
        self.items = items
        self.cash = Currency(gold, silver, copper)

    # transfer chest contents to character
    def loot(self, character):
        character.wallet += self.cash
        self.cash.value = 0,0,0


# main


aragorn = rpg.Character('Aragorn', 'Human', 100, 50, )
galadriel = rpg.Mage('Galadriel', 'Elf', 120, 75, 200)
frodo = rpg.Burglar('Frodo', 'Hobbit', 50, 25)

galadriel.wallet.value = 10, 5, 2
aragorn.wallet.value = 20, 50, 80
frodo.wallet.value = 5, 25, 20


chest = rpg.Chest (['longsword', 'iron helm'], 2, 50, 10)


print (galadriel.__dict__)
chest.loot(galadriel)
print('')
print (galadriel.wallet) 

print (galadriel.__dict__)


galadriel.battle(frodo)
frodo.battle(aragorn)
galadriel.portal('Minas Tirith') 
aragorn.wallet += frodo.wallet 

print (aragorn.wallet)
print (frodo.wallet)

File: javascript/intro/rpg.js
class Currency {
    #gold
    #silver
    #copper

    constructor(gold, silver, copper){
        this.#gold = gold
        this.#silver = silver
        this.#copper = copper
    }

    get value(){
        return [this.#gold, this.#silver, this.#copper] // need to wrap in array as it only returns last value 
    }

    set value(value_array) {
        const [gold, silver, copper] = value_array
        this.#gold = gold
        this.#silver = silver
        this.#copper = copper
    }

    add(gold, silver, copper){
        this.#gold += gold
        this.#silver += silver
        this.#copper += copper
    }

    addCurrency(other) {
        const [g, s, c] = other.value
        this.add(g, s, c)
    }

    // End -user readable representation of the object
    toString() {
        return `${this.#gold}G ${this.#silver}S ${this.copper}C`
    }

}

class Character {
    constructor(name, race, health, attack) {
        this.name = name
        this.race = race
        this.health = health
        this.attack = attack
        this.wallet = new Currency(0, 0, 0)
    
    }

    battle(other) {
        console.log(`${this.name} launches a melee attack at ${other.name}`)
    }
}

class Chest {
    constructor(items, gold, silver, copper) {
        this.items = items
        this.cash = new Currency(gold, silver, copper)
    }

    loot(character) {
        character.wallet.addCurrency(this.cash)
        this.cash.value = [0, 0, 0]
    }

}

// Main

const aragorn = new Character('Aragorn', 'Human', 100, 50, )
const galadriel = new Character('Galadriel', 'Elf', 120, 75, 200)
const frodo = new Character('Frodo', 'Hobbit', 50, 25)

galadriel.wallet.value = [10, 5, 2]
aragorn.wallet.value = [20, 50, 80]
frodo.wallet.value = [5, 25, 20]

chest = new Chest(['longsword', 'iron helm', 2, 55, 10])

console.log(aragorn)
aragorn.battle (galadriel)
console.log('' + frodo.wallet) // Works now because concatenates into the '' string
chest.loot(galadriel)
console.log(galadriel.wallet.value)
console.log(chest.value)